File: python/leetcode/087.py
class Solution(object):
    def isScramble(self, s1, s2):
        from collections import Counter
        if s1 == s2: return True
        if len(s1) != len(s2): return False
        l1, l2 = Counter(s1), Counter(s2)
        if l1 != l2: return False
        n = len(s1)
        for i in range(1, n):
            if self.isScramble(s1[:i], s2[:i]) and self.isScramble(s1[i:], s2[i:]):
                return True
            if self.isScramble(s1[:i], s2[n-i:]) and self.isScramble(s1[i:], s2[:n-i]):
                return True
        return False

File: cpp/leetcode/087.cpp
class Solution {
public:
    bool isScramble(string s1, string s2) {
        if(s1 == s2) return true;
        unordered_map<char, int> map;
        for(auto ch : s1) map[ch]++;
        for(auto ch : s2) map[ch]--;
        for(auto it : map){
            if(it.second != 0) return false;
        }
        int n = s1.size();
        for(int i = 1; i < n; i++){
            if(isScramble(s1.substr(0, i), s2.substr(0, i))  && isScramble(s1.substr(i), s2.substr(i))) return true;
            if(isScramble(s1.substr(0, i), s2.substr(n - i)) && isScramble(s1.substr(i), s2.substr(0, n - i))) return true;
        }
        return false;
    }
};